TUTORIAL.adoc

```diff
@@ -0,0 +1,536 @@
+
+
+== Welcome to Meander
+
+Meander is a term rewriting language embedded inside of Clojure for expressing declarative transforms.
+A language where transformations are pictures; patterns of logical matches and substitution,
+instead of linguistic descriptions built up from function composition.
+The purpose of this tutorial is to explore what programming looks like when we convert between data representations declaratively.
+
+If you just want to try Meander out, hide this tutorial with the `^` button.
+For a guided tour, press the `&gt;` button.
+
+
+== Diving in the deep end
+
+Meander contains a bunch of neat stuff.
+So much so that it may appear confusing as to where to start.
+The link:README.md[readme] starts with the fundamental building blocks and their operation.
+In contrast, this is a top down tutorial.
+We will start at the apex of Meander's power; term rewriting with the `rewrite` macro.
+
+Instead of describing an end to end translation, we will be creating small steps.
+We will write patterns and control how these patterns are applied and reapplied using combinators.
+We will cover the operators available to us inside `rewrite` expressions.
+Finally we shall look at some common situations where we can employ meander to express structural solutions,
+where otherwise we might have used linguistic approaches.
+
+The `rewrite` construct exists in the `strategy` namespace.
+The `strategy` namespace also contains the mechanisms for combining rewrite steps.
+For now it is the only namespace we need concern ourselves with.
+Pretty much anything we could do with the other namespaces, we can achieve with a `rewrite`.
+
+[source,clojure]
+----
+(ns meander.examples
+  (:require [meander.strategy.delta :as m]))
+----
+
+
+== Our first rewrite
+
+[source,clojure]
+----
+((m/rewrite
+  ?x
+  "foo")
+ "bar")
+;=> "foo"
+----
+
+Rewrite takes a match pattern and and a substitution pattern.
+In this case the match pattern is `?x` and the substitution pattern is `"foo"`.
+`?x` is a logic variable. In this pattern it will match with anything, as it stands alone.
+
+Meander's `rewrite` is a macro that constructs a function.
+Meander is compiling an efficient rewrite function for us to use with arbitrary inputs.
+Here we immediately invoke our rewriting function on the input `"bar"`.
+The result of calling the rewrite function on `"bar"` is `"foo"`.
+
+Using the rewrite immediately like this is a convenient way to try our rewrite expressions as we go.
+
+
+== Reusing rewrite
+
+Once we are happy with a transformation, it makes sense to give it a name.
+
+[source,clojure]
+----
+(def foo-bar
+  (m/rewrite
+    ?x
+    "foo"))
+
+(foo-bar "bar")
+;=> "foo"
+(foo-bar "baz")
+;=> "baz"
+----
+
+Meander compiles a function that takes some input, and rewrites it to a new output.
+
+
+== Structural pattern matching
+
+Unlike Clojure's pervasive sequence abstraction that applies to all collections,
+Meander matches structure explicitly.
```

Comment:
An example of the sequence abstraction would be nice, to compare against the declarative approach of meander.